keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,821 @@
+type Node struct {
+    ...
+    Status NodeStatus `json:"status,omitempty" `
+}
+
+```
+
+**Example:**
+
+```
+NodeCapabilities:
+   // Kubelet configurations
+   kubernetes.io/feature/podLevelResources: "true"
+   kubernetes.io/feature/inPlacePodResize: "true"
+   // Runtime configurations 
+   runtime.kubernetes.io/handlers: "runc, runsc"
+   // Kernel capabilities
+   kernel.kubernetes.io/LSMConfig: "seLinux"   
+   // Plugins capabilities
+   csi.kubernetes.io/my-driver/feature/encryptionSupport: "true"
+```
+
+**Note:**  
+
+*   We currently have[ Node Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6279) and[ Node Runtime Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6251) which publish some runtime features through Node Status. They are too narrowly scoped and currently not used for scheduling pods. We can deprecate those fields and introduce them into NodeCapabilities if required. 
+*   Other API options are explored in the [Alternatives](?tab=t.0#heading=h.vfsx3j6ygclj) section. \
+
+
+### Kube-Scheduler Changes
+
+To enable the scheduler to filter nodes based on advertised capabilities, this KEP proposes the introduction of a new [Filter Plugin](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#filter) - `NodeCapabilityFilter`. This plugin will be part of the default scheduler and will be enabled when the `NodeCapabilities` feature gate is active. Additionally, the filter plugin would skip any capability checks for daemonset pods.
+
+`NodeCapabilityFilter` plugin would 
+
+*   Inspect the PodSpec to infer a set of required NodeCapability key-value pairs. This inference is performed using imperative logic built into the scheduler plugin (without new, explicit capability request fields in the PodSpec).
+*   For each node being evaluated, the plugin will access the `node.status.capabilities` map. It will then iterate through the pod's inferred requirements and check if each required capability is present and matches on the node. If any inferred requirement is not met, the plugin will filter the node, making it unschedulable for the pod.
+*   On a mismatch, the plugin will return an FailedScheduling status with a reason that includes the capability that was not found on Nodes.
+
+
+### Admission Controller Changes
+
+To enable the validation of API requests against Node Capabilities, this KEP proposes the introduction of a new admission controller plugin `NodeCapabilityValidator `that will be enabled when the NodeCapabilities feature gate is active. This is necessary because the validation logic that requires cross-referencing objects i,e looking up the `Node` object that a `Pod` is bound to (spec.nodeName). THe admission controller would 
+
+*   Inspects incoming pod update request and infer if it depends on any node capability. 
```

Comment:
We should wait for a concrete use case before adding this. In-place pod resize would have used this, but the timelines don't line up.

Also note that several of the features that could have used this have a separate subresource (`resize` & `ephemeralContainers`)

Reply:
In-place resizing for Guaranteed QoS pods with static CPU policy is an WIP feature that highlights the need for Node Capabilities in admission control. Added more details in the "Example Walkthrough" section.